### ISSUE TITLE
[Redis] Update endpoint returns a HTTP 504 Timeout

### ISSUE THREAD

**Issue by tombuildsstuff** [Azure/azure-rest-api-specs#1499]
👋 

We run nightly acceptance tests against Azure Public in various regions - a selection of these tests are spinning up a Redis Cache instance.

For one of these tests we first create a Redis Cache with RDB Backup Enabled via the API and then disable it by calling the Update endpoint which is returns a `HTTP 504` timeout. Unfortunately this test is flapping a lot, so it's hard to give an exact amount of time where this is broken, however looking at the logs I'd estimate that somewhere in the region of 30-40% of the calls to the Update endpoint return a `HTTP 504` from the server.

The Response we get back from the Go SDK is:

```
redis.GroupClient#Update: Failure responding to request: StatusCode=504 -- Original Error: autorest/azure: Service returned an error. Status=504 Code="GatewayTimeout" Message="The gateway did not receive a response from 'Microsoft.Cache' within the specified time period."
```

However that's merely wrapping the response from the API - which is returning a HTTP 504 - so from what I can tell this is an issue with the Redis API. Would it be possible for someone to take a look?

Thanks!

cc @TimLovellSmith @sarangan12

**Comment by TimLovellSmith** [comment]
@tombuildsstuff does this still happen?

**Comment by tombuildsstuff** [comment]
@TimLovellSmith yes, but we only see this infrequently (I’m more concerned that end users will see it than the odd test failure) - as such I’m wondering if 504 would make a good candidate for the SDK “auto-retry”?

**Comment by tombuildsstuff** [comment]
@TimLovellSmith is there any update on this? ^

In addition - looking at our acceptance tests, we're also see the following connection drops from the Redis API's when attempting to Update the Backup configuration via a HTTP Patch (using the Go SDK):

```
redis.GroupClient#Update: Failure sending request: StatusCode=0 -- Original Error: Patch https://management.azure.com/subscriptions/*******/resourceGroups/acctestRG-2338313560319322468/providers/Microsoft.Cache/Redis/acctestRedis-2338313560319322468?api-version=2016-04-01: http: ContentLength=178 with Body length 0
```

Our CI system has picked this test up as flakey - with 87 failures out of 172 runs (a success rate of 49.4%) - would it be possible for someone to take a look into this?

Thanks!

**Comment by TimLovellSmith** [comment]
@tombuildsstuff Well that is intruiguing. I found the latest request for this cache that failed with 504 and it is because in our backend we repeatedly try to generate SAS tokens for the storage account, which I assume was specified in the request, but fail DNS lookup for it:

Exception: Microsoft.WindowsAzure.Storage.StorageException: The remote name could not be resolved: 'unlikely23exst2acctr6mt.blob.core.windows.net' 

I don't think Azure DNS would be so flakey it could cause your test to fail 50% of the time - is there any chance your test is working with a storage account which isn't fully created yet, or could already have been deleted?

**Comment by TimLovellSmith** [comment]
I suggest this should be closed as not repro, unless further details are provided. Clicking unassign me.
/cc @tombuildsstuff

**Comment by dkirrane** [comment]
I hit this from terraform fairly often. And it even happens from the portal sometimes. For example when I tried to delete Redis from the portal I got this
![image](https://user-images.githubusercontent.com/1280116/122592903-d0961f00-d05c-11eb-8aa0-365ddf652264.png)

**Comment by WolfspiritM** [comment]
There seems to be some problem with the redis api here, too. Getting the state (even without any change) via terraform constantly returns a 504:
```Error: retrieving Redis Cache "redis..." (Resource Group "..."): redis.Client#Get: Failure sending request: StatusCode=504 -- Original Error: context deadline exceeded```

**Comment by dkirrane** [comment]
Hit the same issue again to day creating Redis cluster

```
Error: waiting for the create of Redis Cache "test-XXXX-redis" (Resource Group "test-XXXX-rg"): Future#WaitForCompletion: the number of retries has been exceeded: StatusCode=200
```

And from the Portal I see

![image](https://user-images.githubusercontent.com/1280116/122906751-66c29180-d34a-11eb-853b-ea567f0ed27d.png)

**Comment by bhordupur** [comment]
We are having this issue as well, in 10-15 tries it will work once.

`: redis.Client#ListKeys: Failure sending request: StatusCode=504 -- Original Error: context deadline exceeded`

**Comment by jyotisingh** [comment]
Same issue as [this comment](https://github.com/Azure/azure-rest-api-specs/issues/1499#issuecomment-865892944) for us. Recently we started using `Azure Cache for Redis` and provision the server using terraform scripts. 6 out of 10 times the script fails with the following error:
```
Error: listing keys for Redis Cache <NAME of the resource>: redis.Client#ListKeys: Failure sending request: StatusCode=504 -- Original Error: context deadline exceeded
```
Our provisioning scripts run upon every commit so we have this failure multiple times in a day. Only solution at the moment seems like re-running the pipeline and chances are it may pass, but it fails quite often. 

If you could make a fix for this, then it's great. 
If you have a workaround to offer to get around this issue, even that would  be useful for us. 
Else the only option I see for us is to move to own provisioning of redis-server which would be quite some work for us.

**Comment by jeicob22** [comment]
Similar situation like the one in [this comment](https://github.com/Azure/azure-rest-api-specs/issues/1499#issuecomment-865912254) here.

We don't modify the Redis quite often (max ~1 modification/year)... But this delays all the production infrastructure deployment and IaC testing as it (randomly) fails during the read of the redis clusters status in Azure, causing fails on the terraform plan process.

This is really annoying for us as we can't bypass this without moving "key resources" of our infrastructure out of the IaC manifests, which at the end could evolve into manual modifications and insert human errors in our automated processes... But also seems to be (randomly) failing the GUI in the Azure Portal, returning a 504 error...

Not sure if this helps, but we've observed this fails more often during working hours.
Also some members of the team have observed that the users with more privileges suffer this situation less often, but we don't consider this is a real/relevant fact.

**Comment by jeicob22** [comment]
When I say "[RANDOMLY](https://github.com/Azure/azure-rest-api-specs/issues/1499#issuecomment-868374393)", I want to say that there are some days on which we can't work on the IaC as this is blocking us 100%. Today we've been receiving that 504 error since 5 hours ago without any kind of success.

**Comment by restfulhead** [comment]
Our Terraform deployments started to fail today with a 504 error. We did not even make any changes, just updating the state by listing keys fails: `redis.Client#ListKeys: Failure sending request: StatusCode=504`. Very frustrating

**Comment by WolfspiritM** [comment]
According to azure support they have a "regional management plane issue in West Europe for multiple caches" and are aware of it and their Product Group is working on it

**Comment by rvava** [comment]
Thanks a lot @WolfspiritM, we had the same issue for the past 3 days! After reading your comment, we deployed it on France Central instead of West Europe and it worked at the first attempt ! 
Caution though, prices are not the same regarding the region.

**Comment by qbast** [comment]
This is not just modification - even reading cache status times out so this breaks any terraform module that includes cache resource since even 'refresh' operation cannot be completed
